identity-did/src/verification/verification_method.rs

```diff
@@ -104,37 +152,139 @@ impl<T> VerificationMethod<T> {
+      key_data: self.key_data,
+      properties: self.properties,
+    }
+  }
+
+  /// Fallible version of [`VerificationMethod::map`].
+  pub fn try_map<C, F, E>(self, mut f: F) -> Result<VerificationMethod<C, T>, E>
+  where
+    C: DID,
+    F: FnMut(D) -> Result<C, E>,
+  {
+    Ok(VerificationMethod {
+      id: self.id.try_map(&mut f)?,
+      controller: f(self.controller)?,
+      key_type: self.key_type,
+      key_data: self.key_data,
+      properties: self.properties,
+    })
+  }
 }
 
-impl<T> Display for VerificationMethod<T>
+impl<D, T> VerificationMethod<D, T>
 where
+  D: DID,
+  T: Revocation + Default,
+{
+  // ===========================================================================
+  // Constructors
+  // ===========================================================================
+
+  /// Creates a new [`IotaVerificationMethod`] from the given `did` and public key.
+  pub fn new(did: D, key_type: KeyType, public_key: &PublicKey, fragment: &str) -> Result<Self> {
+    let method_fragment: String = if !fragment.starts_with('#') {
+      format!("#{}", fragment)
+    } else {
+      fragment.to_owned()
+    };
+    let id: DIDUrl<D> = did.to_url().join(method_fragment)?;
+
+    let mut builder: MethodBuilder<D, T> = MethodBuilder::default().id(id).controller(did);
+    match key_type {
+      KeyType::Ed25519 => {
+        builder = builder.key_type(MethodType::Ed25519VerificationKey2018);
+        builder = builder.key_data(MethodData::new_multibase(public_key));
+      }
+    }
+    builder.build()
+  }
+
+  /// Creates a new [`MerkleKeyCollection2021`](MethodType::MerkleKeyCollection2021) method from
+  /// the given key collection.
+  pub fn new_merkle_key<M>(did: D, keys: &KeyCollection, fragment: &str) -> Result<Self>
+  where
+    M: MerkleDigest,
+  {
+    let method_fragment: String = if fragment.starts_with('#') {
+      fragment.to_owned()
+    } else {
+      format!("#{}", fragment)
+    };
+    let id: DIDUrl<D> = did.to_url().join(method_fragment)?;
+
+    MethodBuilder::default()
+      .id(id)
+      .controller(did)
+      .key_type(MethodType::MerkleKeyCollection2021)
+      .key_data(MethodData::new_multibase(&keys.encode_merkle_key::<M>()))
+      .build()
+      .map_err(Into::into)
+  }
+
+  /// Revokes the public key of a Merkle Key Collection at the specified `index`.
+  ///
+  /// # Errors
+  ///
+  /// - if this is called on a `VerificationMethod` with a type other than [`MethodType::MerkleKeyCollection2021`].
+  /// - if the index exceeds u32.
+  pub fn revoke_merkle_key(&mut self, index: usize) -> Result<bool> {
+    if self.key_type() != MethodType::MerkleKeyCollection2021 {
+      return Err(Error::InvalidMethodRevocation);
+    }
+
+    // TODO: map this to a descriptive error?
+    let index: u32 = index.try_into().map_err(|_| Error::InvalidMethodRevocation)?;
+    let mut revocation: BitSet = self.revocation()?.unwrap_or_else(BitSet::new);
+    let revoked: bool = revocation.insert(index);
```

Comment:
Yeah I made the parameter take `u32` directly to eliminate this potential error case. I also found it weird and would consider `u32` a reasonable upper-bound, which might be better since `usize` can differ based on the platform.

Reply:
Resolving this discussion in light of the planned changes to `KeyCollection`.